### PR TITLE
Lpe vision vel

### PR DESCRIPTION
**Pull request by jgoppert**
This add support for vision speed mavlink message and estimation integration with LPE.

### CODE REVIEW

File: src/modules/local_position_estimator/sensors/vision_velocity.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
+#include "../BlockLocalPositionEstimator.hpp"
+#include <systemlib/mavlink_log.h>
+#include <matrix/math.hpp>
+
+extern orb_advert_t mavlink_log_pub;
+
+static const uint32_t 		REQ_VISION_VEL_INIT_COUNT = 10;
+static const uint32_t 		VISION_VELOCITY_TIMEOUT =    500000;	// 0.5 s
```
Comment (jgoppert) [inline src/modules/local_position_estimator/sensors/vision_velocity.cpp]: @ecmnet Have this set to 0.5 sec,  so should be  what you want, maybe we should make it s param, but just worried about mem for smaller boards.

**Comment by ecmnet** [comment]
@jgoppert Does this work in SITL too (javaSIM)? I tried to publish velocity 0,0,0 to lpe (at 10Hz), I also saw it initializing, but the result was like this:

xychart

I expected it to be stable at the takeoff-position instead.

**Comment by jgoppert** [comment]
@ecmnet I haven't tried this yet, there must be a bug, it should be holding position. I'll try to verify the message is actually getting passed through.

**Comment by ecmnet** [comment]
@jgoppert Ok. I tried SITL with some random speeds as well, but I think I have to try that with real vision estimation with my quad then. NED is ok for me, although that is a bit difficult for me as my vision system takes the first estimated position as frame. So I have to capture attitude at that point of time to rotate the vision information to NED.

**Comment by ecmnet** [comment]
@jgoppert I tried your branch with my real device now, but with no success. Providing VISION_SPEED_ESTIMATE messages only, I can see lpe being initialized, but no XY position estimate is published. Providing both VISION_SPEED_ESTIMATE and VISION_POSITION_ESTIMATE leads to published position with infinite values. Could you help me what I am doing wrong here?

**Comment by jgoppert** [comment]
@ecmnet Looking into it. Can you upload a log?

**Comment by ecmnet** [comment]
@jgoppert This was not flying. So I simulated flight with propellers off like this:

xychart


And here's the corresponding log:  http://logs.uaventure.com/view/R4Bccgov9mTKuz6k3FmKvH
Please note, that this was on ground.

**Comment by jgoppert** [comment]
Yeah, velocity  correction isn't happening , have the message log?

**Comment by ecmnet** [comment]
Nothing special here: [MSGS_~98.TXT](https://github.com/PX4/Firmware/files/514563/MSGS_.98.TXT)

**Comment by jgoppert** [comment]
Yeah, was looking at the timeout. Seems it was running the whole time.  Set LPE_VXY_PU B really high, like 2.0 so that it publishes whatever it has. My guess is that it is estimating but not publishing since the covariance is too high.

**Comment by ecmnet** [comment]
ok, i give that a try

**Comment by ecmnet** [comment]
@jgoppert I set LPE_VXY_PUB to its maximum (which is 1). Still no position published.

I additionally forced a re-init of the odometry, which leads to an lpe re-init as well (by causing a timeout):

[lpe] vision velocity init: -0.05 -0.23  0.27 m s
[vis] reset odometry: Odometry speed
[lpe] vision velocity timeout 
[lpe] vision velocity init:  0.00  0.11  0.02 m s
[vis] reset odometry: Quality
[lpe] vision velocity timeout 
[lpe] vision velocity init: -0.03  0.01 -0.01 m s
[vis] reset odometry: Quality
[lpe] vision velocity timeout 
[vis] reset odometry: Quality
[lpe] vision velocity init:  0.01  0.03  0.06 m s
[vis] reset odometry: Quality
[lpe] vision velocity timeout 
[lpe] vision velocity init:  0.00 -0.00  0.00 m s
[vis] reset odometry: IMU.Rot.speed
[lpe] vision velocity timeout 
[lpe] vision velocity init:  0.25 -0.04  0.11 m s

Note, that I am sending velocities only.

**Comment by ecmnet** [comment]
Seems also that my vision speed is quite accurate:

chart

While standing still, my standard deviation is less than 0.1.

**Comment by jgoppert** [comment]
@ecmnet For my flow setup I had similar issues, Try setting LPE_PN_V = 1 and LPE_VXY_PUB = 1, that made my setup a lot more stable.

**Comment by ecmnet** [comment]
@jgoppert Even with those settings, no position estimation is published.

**Comment by jgoppert** [comment]
I've been working on my system the past 3 days, will push some changes to
help you soon and some of my logs.

**Comment by ecmnet** [comment]
Thanks a lot. In the meantimeI have adjusted current master with my visionOrigin-patch and publish position again. Indeed I had the same behavior (no position published) when setting  LPE_XY_PUB too low.

**Comment by ecmnet** [comment]
@jgoppert With mounted flow, I get a similar drift on ground like the baro. Reason are timeouts of flow on ground. Flow itself works fine for me, better than before. Would you mind to publish the flow based position accuracy in PosHold mode as a benchmark?

**Comment by jgoppert** [comment]
Yes, I can publish a log file for position hold. I can probably use vision as a ground truth and disable it. Is your camera focused at infinity? I focused mine at infinity so that it works well at high altitude. I was able to achieve position hold and auto mission with just flow at 8m. Unfortunately, this means you need to be above about 1.5m to get good optical flow performance. I have set the min agl for flow to 0, we probably need to increase this back to around 1m so it doesn't try to turn on while on the ground and out of focus. Alternatively we could play with the required flow quality.

**Comment by ecmnet** [comment]
@jgoppert I think flow quality would not help in all cases as e.g. my flow reports some quality even on ground. On the other hand, for indoor use cases, 1m seems to be too high. Maybe a good idea is to use the distance sensor as criteria. As soon as it reports some valid data, flow might be valid.

**Comment by ecmnet** [comment]
@jgoppert I achieve about +/-8cm with flow as well as with vision currently. It is interesting that the Y-axis-accuracy is twice as bad as the on in X axis direction.

**Comment by jgoppert** [comment]
What is your long term drift rate in cm/min? I don't have an exact number for mine but I've left it in position hold for about 10  minutes without noticeable drift. The y axis oscillation might be control related. Try tuning your roll p gain up/down and see if it is reduced.

**Comment by ecmnet** [comment]
@jgoppert I don't have that figure. But within that accuracy, the vehicle definitely stays stable (I usually test it for 2 mins with power tethered).

**Comment by ecmnet** [comment]
@jgoppert A 60sec flight with PosHold enabled after takeoff and manual stabilization:

xychart

**Comment by jgoppert** [comment]
@ecmnet are you testing with this vision velocity branch for your vision correct, and is it working well now?

Also, it would seem the +/- 10 cm drift appears to just be from the controller. Any drift in the estimate from the true position would not show up in the above plot since we don't have a measurement of ground truth.

**Comment by ecmnet** [comment]
@jgoppert I know, the absolute drift is just visually :-). Meanwhile i tuned the controller a bit and I cut the oscillations by half. Yes, I am using this branch, but I am not publishing velocity but vision position currently. I have added my offset patch again.

**Comment by jgoppert** [comment]
Yeah, I figured you did, just trying to be explicit for others reading. This branch might be a bit behind master, so might consider using that for vision position.

**Comment by jgoppert** [comment]
Too much mem usage for lpe to add now. Idea added to IEKF.

**Pull request closed without merging**